controllers/containerizedworkload_controller_helper.go

```diff
@@ -0,0 +1,158 @@
+package controllers
+
+import (
+	"context"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	oamv1alpha2 "github.com/oam-dev/core-resource-controller/api/v1alpha2"
+)
+
+const (
+	OAMResourceTypeLabel = "oam.dev/type"
+	OAMResourceNameLabel = "oam.dev/name"
+	KindDeployment       = "Deployment"
+	KindService          = "Service"
+)
+
+// OAMResourceTypes
+type OAMResourceTypes string
+
+// Supported OAM Resource Types
+const (
+	workloadType OAMResourceTypes = "workload"
+)
+
+// create a corresponding deployment
+func (r *ContainerizedWorkloadReconciler) renderWorkload(ctx context.Context,
+	workload *oamv1alpha2.ContainerizedWorkload) (*appsv1.Deployment, error) {
+	var RevisionHistoryLimit int32 = 100
+	deployName := workload.Name + "-deployment"
+	depl := appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: appsv1.SchemeGroupVersion.String(),
+			Kind:       KindDeployment,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deployName,
+			Namespace: workload.Namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{OAMResourceTypeLabel: string(workloadType), OAMResourceNameLabel: deployName},
+			},
+			RevisionHistoryLimit: &RevisionHistoryLimit,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{OAMResourceTypeLabel: string(workloadType), OAMResourceNameLabel: deployName},
+				},
+				Spec: corev1.PodSpec{
+					Containers: workload.Spec.Containers,
+				},
+			},
+		},
+	}
+
+	// always set the controller reference so that we know which object owns this.
+	if err := ctrl.SetControllerReference(workload, &depl, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	return &depl, nil
+}
+
+// create a service for the deployment
+func (r *ContainerizedWorkloadReconciler) renderService(ctx context.Context, deploy *appsv1.Deployment,
+	port *corev1.ContainerPort) (*corev1.Service, error) {
+	svc := corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       KindService,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deploy.Name + "-service",
+			Namespace: deploy.Namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       strings.ToLower(string(port.Protocol)),
+					Port:       8080,
```

Comment:
why hardcode `8080` here?

Reply:
what port shall we use? I am not sure if it is defined in the containerizedWorkloads

Reply:
It is a field in `ContainerPort` https://github.com/crossplane/crossplane/blob/b724d2eba282a7fb2cea9e1c9501dbfa33d872b7/apis/oam/v1alpha2/core_workload_types.go#L189

Reply:
Disregard my comment :) I see this is picking a port to expose :+1:

Reply:
exactly, this was missing in our spec and we need to formalize it.